pkg/build/controller/image_change_controller_test.go

```diff
@@ -0,0 +1,281 @@
+		t.Errorf("Image substitutions not properly setup for new build: %s |", buildCreator.imageSubstitutions["registry.com/namespace/imagename"])
+	}
+	if buildConfigUpdater.buildcfg == nil {
+		t.Fatal("Expected buildConfig update when new image was created!")
+	}
+	if buildConfigUpdater.buildcfg.Triggers[0].ImageChange.LastTriggeredImageID != "newImageID123" {
+		t.Errorf("Expected imageID newImageID123, got %s", buildConfigUpdater.buildcfg.Triggers[0].ImageChange.LastTriggeredImageID)
+	}
+}
+
+func TestNewImageIDDefaultTag(t *testing.T) {
+	// valid configuration using default tag, new build should be triggered.
+	buildcfg := mockBuildConfig("registry.com/namespace/imagename", "registry.com/namespace/imagename", "testImageRepo", "")
+	controller := mockImageChangeController(buildcfg, "testImageRepo", "registry.com/namespace/imagename", map[string]string{buildapi.DefaultImageTag: "newImageID123"})
+	controller.HandleImageRepo()
+	buildCreator := controller.BuildCreator.(*mockBuildCreator)
+	buildConfigUpdater := controller.BuildConfigUpdater.(*mockBuildConfigUpdater)
+
+	if buildCreator.buildcfg == nil {
+		t.Error("Expected new build when new image was created!")
+	}
+	if buildCreator.imageSubstitutions["registry.com/namespace/imagename"] != "registry.com/namespace/imagename:newImageID123" {
+		t.Errorf("Image substitutions not properly setup for new build using default tag: %s |", buildCreator.imageSubstitutions["registry.com/namespace/imagename"])
+	}
+	if buildConfigUpdater.buildcfg == nil {
+		t.Fatal("Expected buildConfig update when new image was created!")
+	}
+	if buildConfigUpdater.buildcfg.Triggers[0].ImageChange.LastTriggeredImageID != "newImageID123" {
+		t.Errorf("Expected imageID newImageID123, got %s", buildConfigUpdater.buildcfg.Triggers[0].ImageChange.LastTriggeredImageID)
+	}
+}
+
+func TestNonExistentImageRepository(t *testing.T) {
+	// this buildconfig references a non-existent imagerepo, so an update to the real imagerepo should not
+	// trigger a build here.
+	buildcfg := mockBuildConfig("registry.com/namespace/imagename", "registry.com/namespace/imagename", "testImageRepo", "testTag")
+	controller := mockImageChangeController(buildcfg, "otherImageRepo", "registry.com/namespace/imagename", map[string]string{"testTag": "newImageID123"})
+	controller.HandleImageRepo()
+	buildCreator := controller.BuildCreator.(*mockBuildCreator)
+	buildConfigUpdater := controller.BuildConfigUpdater.(*mockBuildConfigUpdater)
+
+	if buildCreator.buildcfg != nil {
+		t.Error("New build created when a different repository was updated!")
+	}
+	if len(buildCreator.imageSubstitutions) != 0 {
+		t.Errorf("Should not have had any image substitutions since tag does not exist in imagerepo")
+	}
+	if buildConfigUpdater.buildcfg != nil {
+		t.Error("BuildConfig was updated when a different repository was updated!")
+	}
+}
+
+func TestNewImageDifferentTagUpdate(t *testing.T) {
```

Comment:
This particular test does not produce a new build because the tag "testTag" does not exist in the imagerepo at all.  Which is a valid test, but we need another test for the case where both "testTag" and "otherTag" exist in the imageRepo, but "otherTag" is the tag that's being updated.  In that situation, we should not trigger a build either, but the reason we don't trigger a build is different.

In the first case, we don't trigger a build because there's no imageid found in the imagerepo for the tag in question.
In the second case, we don't trigger a build because the imageid for the tag in question has not changed since last time we ran the build.